File: kernel.m

function kernel()
%disp("Hello from kernel")
if ~coder.target("MATLAB")
    coder.cinclude("tomos/msh_lib/mshlib.h")
end
colors={'black' 'blue' 'green' 'cyan' 'red' 'magenta' 'brown' 'light_grey' 'dark_grey' 'light_blue' 'light_green' 'light_cyan' 'light_red' 'light_magenta' 'light_brown' 'white'};
vga_color = colors;

[vga_width, vga_height,terminal_row,terminal_column,terminal_color,terminal_buffer] = terminal_initialize(vga_color);
%[vga_width,vga_height,terminal_row,terminal_column,terminal_color,terminal_buffer]
[terminal_column, terminal_row, terminal_buffer] = terminal_writestring('Hello from matlab kernel!',vga_width,vga_height,terminal_row,terminal_column,terminal_color,terminal_buffer);
end

function ret = vga_entry_color(fg, bg)
%disp("Hello from vga_entry_color")
ret = bitor(fg, bitshift(bg, 4));
ret = uint8(ret);
%whos ret b
end

function ret = vga_entry(uc, color)
%disp("Hello from vga_entry")
%uc = uint8(uc);
%uc = str2ascii(uc,1);
assert(isa(uc, "char"))
%color = uint8(color);
assert(isa(color,"uint8"))
ret = bitor(uint16(uc), bitshift(uint16(color),8));
%ret = uint16(ret);
assert(isa(ret,"uint16"))
end

function [vga_width,vga_height,terminal_row,terminal_column,terminal_color,terminal_buffer] = terminal_initialize(vga_color)
%disp("Hello from terminal_initialize")
vga_width = uint32(80);
vga_height = uint32(25);
terminal_row = uint32(0);
%terminal_column= uint32(0);
terminal_column=double(0);
terminal_color = vga_entry_color(find(vga_color=="light_grey")-1, find(vga_color=="black")-1);%always need -1 with arrays
terminal_color = uint8(terminal_color);
%terminal_buffer = uint16(vga_width * vga_height);
%terminal_buffer = createArray([1 vga_width*vga_height], "uint16");
terminal_buffer = zeros([1 vga_width*vga_height], "uint16");
for y=0:vga_height
    for x=0:vga_width
        if ~coder.target("MATLAB")
            index = (y*vga_width) + x;
            value = vga_entry(' ', terminal_color);
            coder.ceval("use_video_memory",uint32(index),uint16(value));
        end
    end
end

end

function terminal_buffer = terminal_putentryat(c, color, x, y,vga_width,terminal_buffer)
%disp("Hello from terminal_putentryat")
index = uint32(y * vga_width + x + 1);%+1 always needed
%terminal_buffer(x, y) = vga_entry(c, color);
%terminal_buffer(index) = vga_entry(c, color);
value = vga_entry(c, color);
if ~coder.target("MATLAB")
    coder.ceval("use_video_memory",uint32(index),uint16(value));
end
end

function [terminal_column,terminal_row, terminal_buffer] = terminal_putchar(c, vga_width,vga_height,terminal_row,terminal_column,terminal_color,terminal_buffer)
%disp("Hello from terminal_putchar")
terminal_buffer = terminal_putentryat(c, terminal_color, terminal_column, terminal_row,vga_width,terminal_buffer);
if terminal_column+1 == vga_width
    terminal_column = 0;
    if terminal_row+1 == vga_height
        terminal_row = uint32(0);
    else
        terminal_row = terminal_row+1;
    end
else
    %assert(isa(terminal_column,"uint32"))
    %whos terminal_column
    %terminal_columnType = class(terminal_column);
    %terminal_column = cast(terminal_column, "uint32")
    terminal_column = double(terminal_column+1);
    %terminal_column = uint32(terminal_column);
    %terminal_column = cast(terminal_column,"uint32")
    %what the fuck
    %whos terminal_column
end

end

function [terminal_column, terminal_row,terminal_buffer] = terminal_write(data, size,vga_width,vga_height,terminal_row,terminal_column,terminal_color,terminal_buffer)
%disp("Hello from terminal_write")
for i=1 : size
    [terminal_column,terminal_row,terminal_buffer]  = terminal_putchar(data(i),vga_width,vga_height,terminal_row,terminal_column,terminal_color,terminal_buffer);
end
end

function [terminal_column,terminal_row,terminal_buffer] = terminal_writestring(data, vga_width,vga_height,terminal_row,terminal_column,terminal_color,terminal_buffer)
assert(isa(data, "char"))
%[vga_width,vga_height,terminal_row,terminal_column,terminal_color,terminal_buffer]
[terminal_column, terminal_row,terminal_buffer] = terminal_write(data, length(data),vga_width,vga_height,terminal_row,terminal_column,terminal_color,terminal_buffer);
end

File: sysroot/usr/lib/kernel.c
/*
 * Academic License - for use in teaching, academic research, and meeting
 * course requirements at degree granting institutions only.  Not for
 * government, commercial, or other organizational use.
 *
 * kernel.c
 *
 * Code generation for function 'kernel'
 *
 */

/* Include files */
#include "kernel.h"
#include "mshlib.h"

/* Function Definitions */
void kernel(void)
{
  static const char cv[25] = {'H', 'e', 'l', 'l', 'o', ' ', 'f', 'r', 'o',
                              'm', ' ', 'm', 'a', 't', 'l', 'a', 'b', ' ',
                              'k', 'e', 'r', 'n', 'e', 'l', '!'};
  double terminal_column;
  unsigned int terminal_row;
  int x;
  int y;
  /* disp("Hello from kernel") */
  /* disp("Hello from terminal_initialize") */
  /* terminal_column= uint32(0); */
  /* always need -1 with arrays */
  /* terminal_buffer = uint16(vga_width * vga_height); */
  /* terminal_buffer = createArray([1 vga_width*vga_height], "uint16"); */
  for (y = 0; y < 26; y++) {
    for (x = 0; x < 81; x++) {
      /* disp("Hello from vga_entry") */
      /* uc = uint8(uc); */
      /* uc = str2ascii(uc,1); */
      /* color = uint8(color); */
      /* ret = uint16(ret); */
      use_video_memory((unsigned int)y * 80U + (unsigned int)x, 1824);
    }
  }
  /* [vga_width,vga_height,terminal_row,terminal_column,terminal_color,terminal_buffer]
   */
  terminal_row = 0U;
  terminal_column = 0.0;
  /* disp("Hello from terminal_write") */
  for (y = 0; y < 25; y++) {
    double d;
    unsigned long u;
    unsigned int q0;
    unsigned int qY;
    /* disp("Hello from terminal_putchar") */
    /* disp("Hello from terminal_putentryat") */
    /* +1 always needed */
    /* terminal_buffer(x, y) = vga_entry(c, color); */
    /* terminal_buffer(index) = vga_entry(c, color); */
    /* disp("Hello from vga_entry") */
    /* uc = uint8(uc); */
    /* uc = str2ascii(uc,1); */
    /* color = uint8(color); */
    /* ret = uint16(ret); */
    u = terminal_row * 80UL;
    if (u > 4294967295UL) {
      u = 4294967295UL;
    }
    d = (double)(unsigned int)u + terminal_column;
    if (d < 4.294967296E+9) {
      q0 = (unsigned int)d;
    } else {
      q0 = MAX_uint32_T;
    }
    qY = q0 + 1U;
    if (q0 + 1U < q0) {
      qY = MAX_uint32_T;
    }
    use_video_memory(qY, (unsigned short)(cv[y] | 1792));
    if (terminal_column + 1.0 == 80.0) {
      terminal_column = 0.0;
      qY = terminal_row + 1U;
      if (terminal_row + 1U < terminal_row) {
        qY = MAX_uint32_T;
      }
      q0 = terminal_row + 1U;
      if (terminal_row + 1U < terminal_row) {
        q0 = MAX_uint32_T;
      }
      if (qY == 25U) {
        terminal_row = 0U;
      } else {
        terminal_row = q0;
      }
    } else {
      /* assert(isa(terminal_column,"uint32")) */
      /* whos terminal_column */
      /* terminal_columnType = class(terminal_column); */
      /* terminal_column = cast(terminal_column, "uint32") */
      terminal_column++;
      /* terminal_column = uint32(terminal_column); */
      /* terminal_column = cast(terminal_column,"uint32") */
      /* what the fuck */
      /* whos terminal_column */
    }
  }
}

/* End of code generation (kernel.c) */
